tensorflow/tensorboard
# [Review requested, not ready to merge] Setup mocha frontend unit testing

Adds a `mocha_test` macro, which does the following:
- Compile the TS sources using `ts_web_library`
- Setup a vulcanized html binary target using `tensorboard_html_binary`
- Invokes a python web test using mocha_driver.py to 
  1. Find the `_devserver` target for the `tensorboard_html_binary`, using `os.walk`
  2. Launch that process, and connect to it via WebDriver in Chrome at a hardcoded port
  3. Inspects the Chrome console output to determine when the tests pass or fail.

This has a few flaws, in order of severity:
1. Finding the _devserver target by walking the runfiles is somewhat hacky, but I don't think it is really a problem.
2. The hardcoded port is bad, since it means if you run two tests at once, or run a test while something else is serving at 6006, the test will fail (or could even erroneously pass)
  - It seems impossible to tell the devserver at runtime what port to use, so we either need to change the devserver code, or just write our own web server for this purpose
3. Many tests pass when running on their own, but fail when run in parallel with other tests (or when using `bazel test` instead of `bazel run` in general?), with error messages like:
> SEVERE: Failed to monitor filesystem
java.io.IOException: User limit of inotify instances reached or too many open files
        at sun.nio.fs.LinuxWatchService.<init>(LinuxWatchService.java:64)
        at sun.nio.fs.LinuxFileSystem.newWatchService(LinuxFileSystem.java:47)
        at io.bazel.rules.closure.webfiles.server.Metadata$Reloader.monitorFileSystem(Metadata.java:217)
        at io.bazel.rules.closure.webfiles.server.Metadata$Reloader.run(Metadata.java:205)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
        at java.lang.Thread.run(Thread.java:745)

Nonetheless, please take a look at the general approach, and suggest ways to improve it. Also, @jart I'd like your feedback in particular on whether we should make changes to the webfiles devserver, or code up a new web server for this purpose, or take a different approach entirely.

Also, there are a number of tests which are currently failing (unsurprising, since the tests have been inoperative for more than a year). I'll fix those in a subsequent PR.
